garak/generators/ollama.py

```diff
@@ -0,0 +1,83 @@
+"""Ollama interface"""
+
+from typing import List, Union
+
+import backoff
+import ollama
+
+from garak import _config
+from garak.generators.base import Generator
+
+
+def _give_up(error):
+    return isinstance(error, ollama.ResponseError) and error.status_code == 404
+
+
+class OllamaGenerator(Generator):
+    """Interface for Ollama endpoints
+
+    Model names can be passed in short form like "llama2" or specific versions or sizes like "gemma:7b" or "llama2:latest"
+    """
+
+    DEFAULT_PARAMS = Generator.DEFAULT_PARAMS | {
+        "timeout": 30,  # Add a timeout of 30 seconds. Ollama can tend to hang forever on failures, if this is not present
+        "host": "127.0.0.1:11434",  # The default host of an Ollama server. This should maybe be loaded from a config file somewhere
+    }
+
+    active = True
+    generator_family_name = "Ollama"
+
+    def __init__(self, name="", config_root=_config):
+        super().__init__(name, config_root)  # Sets the name and generations
+
+        self.client = ollama.Client(
+            self.DEFAULT_PARAMS["host"], timeout=self.DEFAULT_PARAMS["timeout"]
+        )  # Instantiates the client with the timeout
```

Comment:
This class currently leaves `parallel_capable` set `True`, the `ollama.Client` object utilizes an `RLock` which prevents pickle of the object for multiprocessing when `--parallel_attempts` or `--parallel_requests` is set > 1.

```suggestion
    # avoid attempt to pickle the client attribute
    def __getstate__(self) -> object:
        self._clear_client()
        return dict(self.__dict__)

    # restore the client attribute
    def __setstate__(self, d) -> object:
        self.__dict__.update(d)
        self._load_client()

    def _load_client(self):
        self.client = ollama.Client(
            self.host, timeout=self.timeout
        )  # Instantiates the client with the timeout

    def _clear_client(self):
        self.client = None

    def __init__(self, name="", config_root=_config):
        super().__init__(name, config_root)
        self._load_client()
```

Here is the test configuration resulting in error against a local ollama service:
```
$python -m garak -m ollama -n llama3.1 -p continuation --parallel_requests 5
garak LLM vulnerability scanner v0.9.0.16.post1 ( https://github.com/leondz/garak ) at 2024-09-10T16:35:11.130523
📜 logging to /home/jemartin/.local/share/garak/garak.log
🦜 loading generator: Ollama: llama3.1
📜 reporting to /home/jemartin/.local/share/garak/garak_runs/garak.2281febd-9657-4b77-a5f9-9b734f7922ea.report.jsonl
🕵️  queue of probes: continuation.ContinueSlursReclaimedSlursMiniTraceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main                                                                                                                                      | 0/5 [00:00<?, ?it/s]
  File "<frozen runpy>", line 88, in _run_code
  File "/home/jemartin/Projects/nvidia/garak/garak/__main__.py", line 14, in <module>
    main()
  File "/home/jemartin/Projects/nvidia/garak/garak/__main__.py", line 9, in main
    cli.main(sys.argv[1:])
  File "/home/jemartin/Projects/nvidia/garak/garak/cli.py", line 509, in main
    command.probewise_run(
  File "/home/jemartin/Projects/nvidia/garak/garak/command.py", line 222, in probewise_run
    probewise_h.run(generator, probe_names, evaluator, buffs)
  File "/home/jemartin/Projects/nvidia/garak/garak/harnesses/probewise.py", line 107, in run
    h.run(model, [probe], detectors, evaluator, announce_probe=False)
  File "/home/jemartin/Projects/nvidia/garak/garak/harnesses/base.py", line 108, in run
    attempt_results = probe.probe(model)
                      ^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/Projects/nvidia/garak/garak/probes/base.py", line 219, in probe
    attempts_completed = self._execute_all(attempts_todo)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/Projects/nvidia/garak/garak/probes/base.py", line 197, in _execute_all
    result = self._execute_attempt(this_attempt)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/Projects/nvidia/garak/garak/probes/base.py", line 156, in _execute_attempt
    this_attempt.outputs = self.generator.generate(
                           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/Projects/nvidia/garak/garak/generators/base.py", line 136, in generate
    for result in pool.imap_unordered(
  File "/home/jemartin/miniconda3/envs/garak/lib/python3.12/multiprocessing/pool.py", line 873, in next
    raise value
  File "/home/jemartin/miniconda3/envs/garak/lib/python3.12/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/home/jemartin/miniconda3/envs/garak/lib/python3.12/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/miniconda3/envs/garak/lib/python3.12/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_thread.RLock' object
```

Reply:
I will just set it to false then.
I'm pretty sure the Ollama server also just queues parallel requests, and deals with them one at a time anyway

Reply:
I guess the suggested code is to circumvent the pickling problem?
The thing is I'm pretty sure that the Ollama server just queues requests made in parallel, and evaluates them sequentially, so is it perhaps not better to just set parallel_capable to False?

Reply:
`parallel_capable = False` works for me.